### PR TITLE
Fix Vulkan compiling issues on Windows

### LINKED ISSUE #59519: SPIR-V codegen fails when building on Windows with Vulkan
## To Reproduce

1. Install `VULKAN_SDK` `1.2.176.1`
2. Install `miniconda` `3.8.10`
3. Install build deps `conda install -y astunparse numpy ninja pyyaml mkl mkl-include setuptools cmake cffi typing_extensions future six requests dataclasses`
4. clone commit `1d9c1c` and `cd  pytorch`
5. build wheel:
```pwsh
$env:USE_VULKAN=1
$env:USE_VULKAN_WRAPPER=0
python setup.py bdist_wheel
```

Build errors because `tools` module is missing. It seems `pytorch` is not added to `PYTHONPATH` in [VulkanCodegen.cmake ](https://github.com/pytorch/pytorch/blob/master/cmake/VulkanCodegen.cmake#L58).
```
Traceback (most recent call last):
  File "D:/a/pytorch-etna/pytorch-etna/pytorch/cmake/../aten/src/ATen/gen_vulkan_spv.py", line 8, in <module>
    from tools.codegen.code_template import CodeTemplate
ModuleNotFoundError: No module named 'tools'
CMake Error at cmake/VulkanCodegen.cmake:72 (message):
  Failed to gen spv.h and spv.cpp with precompiled shaders for Vulkan backend
Call Stack (most recent call first):
  caffe2/CMakeLists.txt:6 (include)
``` 

Build summary:
```
-- ******** Summary ********
--   CMake version         : 3.19.6
--   CMake command         : C:/Miniconda/envs/test/Library/bin/cmake.exe
--   System                : Windows
--   C++ compiler          : C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30037/bin/Hostx64/x64/cl.exe
--   C++ compiler version  : 19.29.30037.0
--   CXX flags             : /DWIN32 /D_WINDOWS /GR /EHsc /w /bigobj -DUSE_PTHREADPOOL -openmp:experimental
--   Build type            : Release
--   Compile definitions   : WIN32_LEAN_AND_MEAN;ONNX_ML=1;ONNXIFI_ENABLE_EXT=1
--   CMAKE_PREFIX_PATH     : C:\Miniconda\envs\test\Lib\site-packages
--   CMAKE_INSTALL_PREFIX  : D:/a/pytorch-etna/pytorch-etna/pytorch/torch
--   CMAKE_MODULE_PATH     : D:/a/pytorch-etna/pytorch-etna/pytorch/cmake/Modules;D:/a/pytorch-etna/pytorch-etna/pytorch/cmake/public/../Modules_CUDA_fix
```

## Expected behavior

Build succeeds.

## Environment
```
OS: Microsoft Windows Server 2019 Datacenter
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.19.6
Libc version: N/A

Python version: 3.8 (64-bit runtime)
Python platform: Windows-10-10.0.17763-SP0
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.2
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.2.0           haa95532_296  
[conda] mkl-include               2021.2.0           haa95532_296  
[conda] mkl-service               2.3.0            py38h2bbff1b_1  
[conda] mkl_fft                   1.3.0            py38h277e83a_2  
[conda] mkl_random                1.2.1            py38hf11a4ad_2  
[conda] numpy                     1.20.2           py38ha4e8547_0  
[conda] numpy-base                1.20.2           py38hc2deb75_0
```

This is on a `windows-2019` github action runner.  The equivalent error happens on a `Windows 10 Pro` desktop.

## Additional context

This issue seems to be a subset or specific version of #55879.

cc @malfet @seemethere @walterddr @peterjc123 @mszhanyi @skyline75489 @nbcsm

### PR DESCRIPTION
**Pull request by mantaionut**
PR based on #61431
Fix USE_VULKAN=1 and USE_VULKAN_WRAPPER=0 not compiling on Windows.
Change designated initializers since they require C++20.
Rename Hasher typename since it's not compiling due to https://developercommunity.visualstudio.com/t/1397858

Fixes #59519

### CODE REVIEW

**Comment by mantaionut** [comment]
@malfet can you please review these changes?
Thanks.

File: aten/src/ATen/native/vulkan/api/Descriptor.cpp [hunk 1]
```diff
@@ -41,20 +41,14 @@ DescriptorSet& DescriptorSet::operator=(DescriptorSet&& other) noexcept {
...
-      },
-  });
+  DescriptorSet::ResourceBinding binder;
+  binder.binding_idx = idx; // binding_idx
+  binder.descriptor_type = shader_layout_signature_[idx]; // descriptor_type
+  binder.is_image = false; // is_image
+  binder.resource_info.buffer_info.buffer = buffer.handle(); // buffer
+  binder.resource_info.buffer_info.offset = buffer.mem_offset(); // offset
+  binder.resource_info.buffer_info.range = buffer.mem_range(); // range
+  add_binding(binder);
```
Comment (ezyang) [inline aten/src/ATen/native/vulkan/api/Descriptor.cpp:51]: Consider std::move() here?

**Comment by mantaionut** [comment]
@ezyang I made the change you requested. Do you know who can review my changes?

File: c10/util/flat_hash_map.h [hunk 2]
```diff
@@ -234,12 +234,14 @@ template <
     typename T,
     typename FindKey,
     typename ArgumentHash,
-    typename Hasher,
+    typename DetailHasher,
```
Comment (SS-JIA) [inline c10/util/flat_hash_map.h:237]: Is this change necessary? Seems that it's simply changing the name of `Hasher` to `DetailHasher`. I believe this file was copied over from [this repo](https://github.com/skarupke/flat_hash_map) so I think we should avoid making changes unless necessary.

**Review by SS-JIA — DISMISSED** [review]
Overall looks good! But I would suggest removing the changes in `c10/util/flat_hash_map.h` since it doesn't seem to be needed.

File: c10/util/flat_hash_map.h (hunk 2 above)
Comment (malfet) [reply c10/util/flat_hash_map.h:237]: It's mentioned in PR description, isn't it
> Rename Hasher typename since it's not compiling due to https://developercommunity.visualstudio.com/t/1397858

**Review by ezyang — APPROVED** [review]
